### PR TITLE
Fixing the DA confusion network & creating a complete test suite for it.

### PR DESCRIPTION
**Pull request by ticcky**
I tried to fix the errors Filip reported during the weekend, cover the whole DialogueActConfusionNetwork class in tests, and clean it up.

### CODE REVIEW

File: alex/components/slu/common.py [hunk 1]
```diff
@@ -44,724 +44,3 @@ def slu_factory(cfg, slu_type=None):
         return slu
 
     raise SLUException('Unsupported SLU parser: %s' % slu_type)
-
```
Comment (tuetschek) [inline alex/components/slu/common.py]: Why are you removing this stuff? Is the change related to the bugs at all?

Comment (ticcky) [inline alex/components/slu/common.py]: Kind of. It uses used some of the methods of da.py which were deleted.

Comment (tuetschek) [inline alex/components/slu/common.py]: OK.

**Comment by jurcicek** [comment]
Still the same problem:

2015-03-23--13-24-16.620791-CET  MainProcess : INFO
    #991    Time: 2015-03-23-13-24-16.619203-CET From: VAD        To: HUB        Command: speech_end(fname="vad-2015-03-23--13-24-14.317038.wav")

2015-03-23--13-24-16.670588-CET  ASR-5      : DEBUG
    ASR: speech_end(fname="vad-2015-03-23--13-24-14.317038.wav")

2015-03-23--13-24-16.674177-CET  ASR-5      : DEBUG
    ASR Hypothesis
    ------------------------------------------------------------
    0.979 CHTĚL BYCH JET Z ANDĚLA NA MALOSTRANSKÉ NÁMĚSTÍ
    0.021 CHTĚL BYCH JET Z ANDĚLA NA MALOSTRANSKÉ NÁMĚSTÍ _NOISE_


2015-03-23--13-24-16.676391-CET  MainProcess : INFO
    #996    Time: 2015-03-23-13-24-16.671891-CET From: ASR        To: HUB        Command: asr_end(fname="vad-2015-03-23--13-24-14.317038.wav")

2015-03-23--13-24-16.800635-CET  SLU-6      : EXCEPTION
    Uncaught exception in SLU process.
    Traceback (most recent call last):
      File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py", line 152, in run
        self.read_asr_hypotheses_write_slu_hypotheses()
      File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py", line 96, in read_asr_hypotheses_write_slu_hypotheses
        slu_hyp = self.slu.parse(data_asr.hyp)
      File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/base.py", line 390, in parse
        return self.parse_nblist(obs, *args, **kwargs)
      File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py", line 815, in parse_nblist
        return self.parse_X(utterance_list, verbose)
      File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py", line 782, in parse_X
        da_confnet.add(p[0][1], self.parsed_classifiers[clser])
      File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/ml/hypothesis.py", line 209, in add
        raise ConfusionNetworkException("Cannot add facts already in the network.")
    ConfusionNetworkException: Cannot add facts already in the network.

Received close event in: DM-7
Process SLU-6:

Received close event in: TTS-9
Traceback (most recent call last):
Received close event in: ASR-5
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py", line 152, in run
Received close event in: MainProcess
    self.read_asr_hypotheses_write_slu_hypotheses()
  File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py", line 96, in read_asr_hypotheses_write_slu_hypotheses
Received close event in: NLG-8
Received close event in: SessionLogger-2
    slu_hyp = self.slu.parse(data_asr.hyp)
  File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/base.py", line 390, in parse
    return self.parse_nblist(obs, *args, **kwargs)
  File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py", line 815, in parse_nblist
    return self.parse_X(utterance_list, verbose)
  File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py", line 782, in parse_X
    da_confnet.add(p[0][1], self.parsed_classifiers[clser])
  File "/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/ml/hypothesis.py", line 209, in add
Received close event in: VoipIO-3
    raise ConfusionNetworkException("Cannot add facts already in the network.")
ConfusionNetworkException: Cannot add facts already in the network.
Received close event in: VAD-4
~/UFAL-DSG-alex-dev/alex/applications/PublicTransportInfoCS
jurcicek@kronos-dev:fixing_da:~/UFAL-DSG-alex-dev/alex/applications/PublicTransportInfoCS$

**Comment by jurcicek** [comment]
I suggest to test the system by calling it.

**Comment by ticcky** [comment]
Thanks, you sent me a different problem. I'll take a look at this one.

**Comment by ticcky** [comment]
I fixed it, added some tests, and tested on spoken dialog. Sorry for the bad commit message. But now it should work now.

**Comment by jurcicek** [comment]
There is still a problem:

2015-03-24--16-13-34.273080-CET  ASR-5      : DEBUG
    ASR: speech_end(fname="vad-2015-03-24--16-13-32.532219.wav")

2015-03-24--16-13-34.277317-CET  ASR-5      : DEBUG
    ASR Hypothesis
    ------------------------------------------------------------
    0.170 A POMOCT CHCI JET ZE
    0.063 A POMOCT CHCI JET _EHM_HMM_
    0.061 A POMOCT JET ZE
    0.041 A TO DOST CHCI JET ZE
    0.038 A POMOCT CHCI JET
    0.031 A TO ZASTÁVCE _NOISE_
    0.025 A POMOCT JET _EHM_HMM_
    0.024 A TO ZASTÁVKY _NOISE_
    0.023 A POMOCT _NOISE_ CHCI JET ZE
    0.023 A POMOCT _NOISE_
    0.022 A TO ZASTÁVKY _EHM_HMM_
    0.022 A POMOCT CHCI JET _NOISE_
    0.016 A TO ZASTÁVKY JET ZE
    0.016 A TO DOST JET ZE
    0.015 A TO DOST CHCI JET _EHM_HMM_
    0.014 A POMOCT _NOISE_ _EHM_HMM_
    0.014 A POMOCT JET
    0.013 A TO MOST CHCI JET ZE
    0.012 A POMOCT TĚ
    0.012 A POMOCT ALEX
    0.012 A POMOCT TAK TO JE


2015-03-24--16-13-37.024692-CET  DM-7       : DEBUG
    DM Output
    ------------------------------------------------------------
    silence()


2015-03-24--16-13-37.027292-CET  MainProcess : INFO
    #1225   Time: 2015-03-24-16-13-37.009677-CET From: DM         To: HUB
     DA: silence()

2015-03-24--16-13-39.740466-CET  SLU-6      : EXCEPTION
    Uncaught exception in SLU process.
    Traceback (most recent call last):
      File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py",
line 152, in run
        self.read_asr_hypotheses_write_slu_hypotheses()
      File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py",
line 96, in read_asr_hypotheses_write_slu_hypotheses
        slu_hyp = self.slu.parse(data_asr.hyp)
      File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/base.py",
line 391, in parse
        return self.parse_nblist(obs, *args, **kwargs)
      File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py",
line 823, in parse_nblist
        return self.parse_X(utterance_list, verbose)
      File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py",
line 792, in parse_X
        da_confnet.sort().merge().prune()
      File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/da.py",
line 1029, in prune
        self.remove(dai)
      File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/ml/hypothesis.py",
line 231, in remove
        self.cn.remove(self.cn[fact_ndx])
    IndexError: list index out of range


Received close event in: NLG-8
Received close event in: MainProcess
Received close event in: DM-7
Received close event in: ASR-5
Process SLU-6:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in
_bootstrap
Received close event in: VoipIO-3
    self.run()
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py",
line 152, in run
    self.read_asr_hypotheses_write_slu_hypotheses()
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py",
line 96, in read_asr_hypotheses_write_slu_hypotheses
    slu_hyp = self.slu.parse(data_asr.hyp)
 Received close event in: SessionLogger-2
 File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/base.py",
line 391, in parse
    return self.parse_nblist(obs, *args, **kwargs)
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py",
line 823, in parse_nblist
    return self.parse_X(utterance_list, verbose)
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py",
line 792, in parse_X
    da_confnet.sort().merge().prune()
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/da.py",
line 1029, in prune
    self.remove(dai)
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/ml/hypothesis.py",
line 231, in remove
Received close event in: VAD-4
    self.cn.remove(self.cn[fact_ndx])
IndexError: list index out of range
Received close event in: TTS-9
~/UFAL-DSG-alex-dev/alex/applications/PublicTransportInfoCS



Best regards,
Filip Jurcicek

-------------------------------------------------------
Work tel. (CZ): +420221914402
Personal tel. (CZ): +420777805048
Skype: bozskyfilip

http://ufal.mff.cuni.cz/filip-jurcicek

**Comment by jurcicek** [comment]
2015-03-24--16-16-38.413615-CET  MainProcess : INFO
    #1460   Time: 2015-03-24-16-16-38.348441-CET From: VAD        To: HUB
     Command: speech_end(fname="vad-2015-03-24--16-16-36.108967.wav")

2015-03-24--16-16-38.417617-CET  ASR-5      : DEBUG
    ASR Hypothesis
    ------------------------------------------------------------
    0.958 CHTĚL BYCH JET Z ANDĚLA NA MALOSTRANSKÉ NÁMĚSTÍ
    0.021 CHTĚL BYCH JET ANDĚLA NA MALOSTRANSKÉ NÁMĚSTÍ
    0.021 CHTĚL BYCH JET Z ANDĚLA NA MALOSTRANSKÉ NÁMĚSTÍ _NOISE_
    0.000 CHTĚL BYCH JET ANDĚLA NA MALOSTRANSKÉ NÁMĚSTÍ _NOISE_


2015-03-24--16-16-38.420194-CET  MainProcess : INFO
    #1465   Time: 2015-03-24-16-16-38.412023-CET From: ASR        To: HUB
     Command: asr_end(fname="vad-2015-03-24--16-16-36.108967.wav")

Received close event in: NLG-8
Received close event in: VoipIO-3
Received close event in: MainProcess
Process SLU-6:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in
_bootstrap
    self.run()
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py",
line 152, in run
    self.read_asr_hypotheses_write_slu_hypotheses()
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/hub/slu.py",
line 96, in read_asr_hypotheses_write_slu_hypotheses
    slu_hyp = self.slu.parse(data_asr.hyp)
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/base.py",
line 391, in parse
    return self.parse_nblist(obs, *args, **kwargs)
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py",
line 823, in parse_nblist
    return self.parse_X(utterance_list, verbose)
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/dailrclassifier.py",
line 792, in parse_X
    da_confnet.sort().merge().prune()
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/components/slu/da.py",
line 1029, in prune
    self.remove(dai)
  File
"/ha/projects/vystadial/work/jurcicek/github/UFAL-DSG-alex-dev/alex/ml/hypothesis.py",
line 231, in remove
    self.cn.remove(self.cn[fact_ndx])
IndexError: list index out of range
Received close event in: SessionLogger-2
Received close event in: ASR-5
Received close event in: DM-7
Received close event in: VAD-4
Received close event in: TTS-9
~/UFAL-DSG-alex-dev/alex/applications/PublicTransportInfoCS


Best regards,
Filip Jurcicek

-------------------------------------------------------
Work tel. (CZ): +420221914402
Personal tel. (CZ): +420777805048
Skype: bozskyfilip

http://ufal.mff.cuni.cz/filip-jurcicek

**Comment by ticcky** [comment]
I'm sorry I tested the system live with HDC SLU instead of the LR SLU. Now it seems to work fine.

**Pull request merged**